### PR TITLE
[MM-62] Update plugin with respect to phase 1 upgrades

### PR DESCRIPTION
**Pull request by ayusht2810**
#### Summary
Updated plugin with respect to phase 1 upgrades

- Updated go version from `1.12` to `1.19`
- Fixed build issues being created from dockerfile
- Updated `build` folder to sync with starter template
- Updated package in go modules to use https://pkg.go.dev/github.com/mattermost/mattermost/server/public
- Updated Makefile
- Updated version in `nvmrc` file and update `package-lock.json` file.
- Added base setup for test cases
- Fix lint errors
- Updated developer docs link in the readme file 

#### Ticket Link

### CODE REVIEW

**Comment by VincentSC** [comment]
Feedback. Still did not get it working, but it's already a big improvement on what is there. See also #206 

The first set of errors could be solved by updating the first line of Dockerfile to `FROM golang:1.18`. This is in line with what `go.mod` defines.

I got this error:
```bash
npm WARN tarball tarball data for redux-offline@git+ssh://git@github.com/enahum/redux-offline.git#885024de96b6ec73650c340c8928066585c413df (sha512-srmJ1vWm8ZQTYflZCf7oUs3WBX83GyCIzsFUpwxUg2wcDHngSHjjShRTCgmkciPkVmM4aJ33i9baYS9jRC+zLA==) seems to be corrupted.
npm ERR! code EINTEGRITY
npm ERR! sha512-srmJ1vWm8ZQTYflZCf7oUs3WBX83GyCIzsFUpwxUg2wcDHngSHjjShRTCgmkciPkVmM4aJ33i9baYS9jRC+zLA== integrity checksum failed when using sha512: wanted sha512-srmJ1vWm8ZQTYflZCf7oUs3WBX83GyCIzsFUpwxUg2wcDHngSHjjShRTCgmkciPkVmM4aJ33i9baYS9jRC+zLA== but got sha512-SThdHzd9uoZ4Uo6I1X4fc1Q1aU5erk6TRZdRVdRtkHrIaYzVhcNqZYGg/euh+LVfiAWQIo6WPeKuGgrZVghBSw==. (49960 bytes)
```
This could be solved by removing [package-lock.json](https://github.com/mattermost/mattermost-plugin-jitsi/blob/master/webapp/package-lock.json).

```
Running golangci-lint
golangci-lint run ./...
ERRO Running error: no such linter "revive"       
make: *** [Makefile:52: check-style] Error 3
```
Needs to be solved in `Makefile`, where a really old version is used. Put it to `v1.55.2`. And that triggered something, I think. :)

```
golangci-lint run ./...
server/api.go:89:23: p.API undefined (type *Plugin has no field or method API) (typecheck)
	bundlePath, err := p.API.GetBundlePath()
	                     ^
server/api.go:160:20: p.API undefined (type *Plugin has no field or method API) (typecheck)
	user, appErr := p.API.GetUser(userID)
	                  ^
server/api.go:190:17: p.API undefined (type *Plugin has no field or method API) (typecheck)
	if _, err := p.API.GetChannelMember(channelID, userID); err != nil {
	               ^
server/command_test.go:30:4: p.SetAPI undefined (type Plugin has no field or method SetAPI) (typecheck)
	p.SetAPI(&apiMock)
	  ^
server/command_test.go:32:39: p.API undefined (type Plugin has no field or method API) (typecheck)
	i18nBundle, err := i18n.InitBundle(p.API, filepath.Join("assets", "i18n"))
	                                     ^
server/command_test.go:119:6: p.SetAPI undefined (type Plugin has no field or method SetAPI) (typecheck)
			p.SetAPI(&apiMock)
			  ^
server/command_test.go:124:41: p.API undefined (type Plugin has no field or method API) (typecheck)
			i18nBundle, err := i18n.InitBundle(p.API, filepath.Join("assets", "i18n"))
			                                     ^
server/command_test.go:157:5: p.SetAPI undefined (type Plugin has no field or method SetAPI) (typecheck)
		p.SetAPI(&apiMock)
		  ^
server/command_test.go:164:40: p.API undefined (type Plugin has no field or method API) (typecheck)
		i18nBundle, err := i18n.InitBundle(p.API, filepath.Join("assets", "i18n"))
		                                     ^
make: *** [Makefile:52: check-style] Error 1
```

**Review by larkox — APPROVED** [review]
Codewise, LGTM.

**Comment by VincentSC** [comment]
Building needs Node 1.19 or higher, else it will stop on:
> /go/pkg/mod/google.golang.org/grpc@v1.60.0/server.go:2161:14: undefined: atomic.Int64

Also the lint-errors found with newer versions of `golangci-lint` need attention. Depending on the used version, different problems arise. This is caused by a change in `go.mod`:
> + google.golang.org/grpc v1.60.0 // indirect

Even with that disabled, it fails on the _existing_ tests of `./webapp/src/components/post_type_jitsi`
Here's one example:
```
 FAIL  src/components/conference/conference.test.tsx (5.437 s)
  Conference
    ✓ should render null if the post type is null (9 ms)
    ✕ should render and initialize the conference interface (17 ms)
    ✓ should execute the hangup command, wait and call the action to close the meeting, and reset the state on closed (488 ms)
    should show the correct buttons depending on the state
      ✕ should have down, open outside, maximize and close buttons (8 ms)
      ✕ should have up, open outside, maximize and close buttons (7 ms)
      ✕ should have open outside, minimize and close buttons (7 ms)
    should show the the loading spinner depending on the state
      ✕ should show loading (7 ms)
      ✕ should not show loading (6 ms)
    should maximize based on the state
      ✓ should toggle tile only if was open before minimize and now is closed (8 ms)
      ✓ should toggle filmstrip only if was open before minimize and now is closed (6 ms)
    should minimize based on the state
      ✓ should toggle tile only if is open before minimizing (3 ms)
      ✓ should toggle filmstrip only if is open before minimize (4 ms)
```

**Comment by VincentSC** [comment]
When updating the Mattermost API to V6, I get:
```
# github.com/mattermost/mattermost-server/v6/shared/mlog
/root/go/pkg/mod/github.com/mattermost/mattermost-server/v6@v6.7.2/shared/mlog/mlog.go:87:11: cannot use generic function logr.Int without instantiation
/root/go/pkg/mod/github.com/mattermost/mattermost-server/v6@v6.7.2/shared/mlog/mlog.go:96:12: cannot use generic function logr.Uint without instantiation
/root/go/pkg/mod/github.com/mattermost/mattermost-server/v6@v6.7.2/shared/mlog/mlog.go:105:14: cannot use generic function logr.String without instantiation
/root/go/pkg/mod/github.com/mattermost/mattermost-server/v6@v6.7.2/shared/mlog/mlog.go:130:12: cannot use generic function logr.Bool without instantiation
/root/go/pkg/mod/github.com/mattermost/mattermost-server/v6@v6.7.2/shared/mlog/mlog.go:143:13: cannot use generic function logr.Array without instantiation
/root/go/pkg/mod/github.com/mattermost/mattermost-server/v6@v6.7.2/shared/mlog/mlog.go:146:11: cannot use generic function logr.Map without instantiation
FAIL	github.com/mattermost/mattermost-plugin-jitsi/server [build failed]
FAIL
make: *** [Makefile:193: test] Error 1
```

**Comment by mickmister** [comment]
For some reason the CI checks are not running for this PR. This should catch the failing webapp test. @ayusht2810 Can you please run the ci commands locally while we figure out the issues with CI? Or have you run them and are not seeing the issues pointed out above?

@toninis Any thoughts on why CI may not be running here?

---

> When updating the Mattermost API to V6 (or higher)

@VincentSC What were your steps for changing the dependency (update paths in source etc.)? This PR already updates the dependency to `github.com/mattermost/mattermost/server/public v0.0.12` which is the recommended way to import the server dependencies now. 

> npm WARN tarball tarball data for redux-offline@git+ssh://git@github.com/enahum/redux-offline.git...

@VincentSC Is your git environment set up to allow for using ssh with GitHub?

> The first set of errors could be solved by updating the first line of Dockerfile to FROM golang:1.18. This is in line with what go.mod defines.

I'm not sure where that `Dockerfile` came from. Looks like a contributor added that early on to help compile without installing Go. @ayusht2810 We should either remove the related docker files in the root dir of the repo or update the Go version in the Dockerfile. I propose removing them if we don't need them, as it's just another piece we need to remember update in this kind of PR that can be missed.

**Comment by VincentSC** [comment]
The Dockerfile is superhandy! It makes sure that an environment is as expected and triaging bugs is not caused by version-problems anywhere. So please don't remove!!

**Comment by VincentSC** [comment]
> @VincentSC What were your steps for changing the dependency (update paths in source etc.)? This PR already updates the dependency to `github.com/mattermost/mattermost/server/public v0.0.12` which is the recommended way to import the server dependencies now.

Then the project uses both and needs a good cleanup. 

> 
> > npm WARN tarball tarball data for redux-offline@git+ssh://[git@github.com](mailto:git@github.com)/enahum/redux-offline.git...
> 
> @VincentSC Is your git environment set up to allow for using ssh with GitHub?

It is done via the docker-build. But this specific environment is indeed not set up for ssh, as I set up something temporary on a server here. Seemingly that is required and not documented? It's weird, as it does not *need* to be done this way, with https being available.

> I propose removing them if we don't need them, as it's just another piece we need to remember update in this kind of PR that can be missed.

My life should not about a single project that defines how my environment should look like. Docker solves that. We have some improvements for just the docker-build, by the way.

**Comment by ayusht2810** [comment]
@VincentSC @mickmister I have fixed the build issues in the Docker file, and now we are able to create a build using the Docker file. Also, regarding the failing of webapp unit test cases, they are passing on my local.

**Comment by VincentSC** [comment]
> I have fixed the build issues in the Docker file, and now we are able to create a build using the Docker file. 
Great! Thanks! I'll test later today.

Does it still need ssh-access to Gitlab or will https suffice?

> Also, regarding the failing of webapp unit test cases, they are passing on my local.
Do you mean on your machine, but within Docker?

**Comment by ayusht2810** [comment]
> Does it still need ssh-access to Gitlab or will https suffice?
@VincentSC I haven't tested that part.

> Do you mean on your machine, but within Docker?
They were passing on my machine as well as within the Docker.

**Comment by VincentSC** [comment]
Yes, it works!

File: docker-make [hunk 1]
```diff
@@ -3,4 +3,4 @@
 set -ex
 
 sudo docker build -t mattermost-plugin-jitsi-builder .
-sudo docker run --rm -it -v $(pwd):/src -w /src mattermost-plugin-jitsi-builder make "$@"
\ No newline at end of file
+sudo docker run --rm -it -v $(pwd):/src -w /src mattermost-plugin-jitsi-builder make "$@"
```
Comment (mickmister) [inline docker-make:6]: It may make sense to have this in `build/custom.mk` to make it into a `make` command, like `make build-docker`. 0/5

**Review by mickmister — APPROVED** [review]

**Comment by VincentSC** [comment]
Merged! I'm happy about this progress, as it enables a lot!

One small problem I just discovered: version is still 2.0.0.

**Pull request merged**